proposals/0012-introduce-reactNativeMeta.md

```diff
@@ -0,0 +1,148 @@
+---
+title: Introducing `reactNativeMeta` to `package.json`, for RN specific metadata
+author:
+  - Lorenzo Sciandra
+date: 25-01-2021
+---
+
+# RFC0012: Introducing `reactNativeMeta` to `package.json`, for RN specific metadata
+
+## Summary
+
+This RFC wants to introduce and formally discuss a new section for the package.json of a react-native project (app or library), called `reactNativeMeta`.
+
+This new section will allow developer to express certain characteristics of their code in a more formalized manner, that can easily used by tooling to act towards the code accordingly.
+
+**Note well**: this is but the first draft of the actual shape and fields. The goal of this draft is to collect feedback and help finalize the spec for this section.
+
+## Basic examples
+
+This is how this section would look like...
+
+...for a library:
+
+```json
+{
+    "name": "@rnx-kit/metro-serializer",
+    "version": "1.0.11",
+    ...,
+    "reactNativeMeta": {
+        "version": "2.3",
```

Comment:
Most package.json extensions for configuration I have seen are not versioned. When I initially read this, I couldn't tell it was meant to be a schema version until I looked lower.

Easier said than done... but would it be sufficient to try to avoid breaking existing fields, and just being additive with new ones? There are definitely projects with more complex configs which have managed that over multiple years.

Reply:
given how I'd expect a good portion of the fields to be related to the new architecture, and how things might vary (think of the codegen) between each version, I think it's going to be very likely that things will change in a "breaking change" way (reason why new arch is still experimental) - and also I can see a future in which a new major version will have removed all references to the new arch as new arch but simply as the "arch we use"

Reply:
Could you give an example of the sort of breaking change you'd want to make?

Reply:
Clarified here: fe3e609f94490d9ac50213f5a16f71764c675d8a

> Could you give an example of the sort of breaking change you'd want to make at the schema level?

As long as we default of using objects all over, we'll be fine extending as we see fit. But I can see how a developer adds a `"mycapability": true` which then wants to extend with more metadata, and this is a breaking change.